README.md
# Py Cortex Intelligence 

## Release Notes
Need to consult the history of our project? [Click Here](CHANGELOG.md)

## How to build locale
```shell
pip install wheel
python setup.py bdist_wheel
```

## How to update on PIP
```
python -m twine upload  dist/*
```

## Cases of Use
### DataInput Parameters

```dictionary
data_input_parameters = {
    'ignoreValidationErrors': True/False,
}
// If you send this parameter, the datainput will ignore errors.
```
### Available Origins

```dictionary
execution_parameters = {
  'name': 'Name of your Integrations',
  'origin': 'Connector',
}

// If you send this to Execution your data need contain
// Dates in YYYY-MM-DD HH:MM:SS format
// Float numbers in XXX.XXX.XXX,YY format
// 'origin' are optional parameters, if you do not send the platform you will try to guess the formats
```

### If you need upload a file to Cortex Application
```python
from pycortexintelligence import functions as cortexfunctions

# Execution Parameters
# You can define Origin, to inform plataform a bundle of parses.
execution_parameters = {
    'name': 'LoadManager PyCortex',
    # 'origin': 'Connector',
}

# DataInput Parameters
data_input_parameters = {
    'ignoreValidationErrors': True,
}

# Loadmanager
loadmanager = 'https://api.cortex-intelligence.com'

# Timeouts
# You can set timeouts for the platform according to the size of the uploaded files
# or use the default
timeout = {
    'file': 300,
    'execution': 600,
}

# DataFormat are Optionally defined
dafault_data_format = {
    "charset": "UTF-8",
    "quote": "\"",
    "escape": "\\",
    "delimiter": ",",
    "fileType": "CSV"
}

# Upload to Cortex
cortexfunctions.upload_to_cortex(
    cubo_id='',
    file_path='',
    plataform_url='CLIENT.cortex-intelligence.com',
    username='',
    password='',
    data_format=dafault_data_format,
    timeout=timeout,
    loadmanager=loadmanager,
    execution_parameters=execution_parameters,
)
```

### If you need download file from Cortex Application
```python
from pycortexintelligence import functions as cortexfunctions

# DataFormat are Optionally defined
dafault_data_format = {
    "charset": "UTF-8",
    "quote": "\"",
    "escape": "\/\/",
    "delimiter": ",",
}

# Select the headers from file
columns = ['Name of Column A', 'Name of Column B']

# OPTIONAL Filters
filters = [
    ['Name of Column A', 'Value'],
    ['Name of Column A', 'Value1|Value2|Value3'],
    ['Name of Column B', 'dd/mm/YYYY'],
    ['Name of Column B', 'dd/mm/YYYY-dd/mm/YYYY'],
]

# Download from Cortex
cortexfunctions.download_from_cortex(
    cubo_id='',
    file_path='',
    plataform_url='CLIENT.cortex-intelligence.com',
    username='',
    password='',
    data_format=dafault_data_format,
    columns=columns,
    filters=filters,
)
```

## CLI Usage
```bash
cortex.py --help
```

### Examples

```bash
cortex.py startproject --name "Project Name" --sname safe_project_name
```

## How to Contribute

### Developers

Developers can access our development manual by [clicking here](CONTRIBUTING.md).

### Community

You can create a new Issue [clicking here](issues/new/choose), and we will start a description about the reported Bug or Feature. 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).

## [Unreleased]

### Added
- CLI to Download Cube/Form from Cortex
- Release 0.1.0 as a Branch
- Releases as a Branch

### Changed
- Improve error handling of functions
- Adjust standardization of function nomenclature

## [0.0.16] - 2021-04-16

### Added
- Support for paramaters on datainput in upload function. View the docs.

### Fixed
- Scape parameter correct.

## [0.0.15] - 2021-03-11

### Added
- Support for paramaters on execution in upload function. View the docs.
```
execution_parameters = {'name': 'Name', 'origin': 'Origin is optional, view the docs'}
```

## [0.0.14] - 2021-02-26

### Added
- Support for custom loadmanagers in upload function. View the docs.
```
loadmanager = "https://api.cortex-intelligence.com"
```

## [0.0.13] - 2021-02-23

### Added
- Support for timeouts in upload function. View the docs.
```
timeout = {
    'file': 300,
    'execution': 600,
}
```

## [0.0.12] - 2021-01-13

### Added
- Support to multiple filters on Download Cube.

## [0.0.10] - 2020-12-22

### Changed
- On default project, dont delete the log file.

## [0.0.9] - 2020-12-14

### Added

- adding developer guidelines to contributions
- adding CHANGELOG.md file
- adding support to download file from Cortex Application

### Added

- create core/messages.py file to index messages

## [0.0.8] - 2020-10-30

### Changed

- using logging module instead of print()
- update on file creation to `utf-8`
- update README.md

### Added

- create core/messages.py file to index messages

## [0.0.7] - 2020-10-30

### Added

- creating a CLI for pycortexintelligence

```bash
cortex.py --help
```

## [0.0.6] - 2020-10-27

### Fixed

- fix bug with calling _upload_local_2_cube_ function.

## [0.0.5] - 2020-10-26

### Added

- added support to data_format on upload_to_cortex

```python
dafault_data_format = {
    "charset": "UTF-8",
    "quote": "\"",
    "escape": "\/\/",
    "delimiter": ",",
    "fileType": "CSV"
}
```

## [0.0.4] - 2020-10-13

### Added

- Create README.md

## [0.0.3] - 2020-10-13

### Added

- Using format to string concatenation.

## [0.0.2] - 2020-10-08

### Added

- Ajust requirements.txt file and configuration to [Pypi](https://pypi.org/)

## [0.0.1] - 2020-10-08

### Added

- Birth of the project in order to simplify the use of Cortex's APIs.

# TODO Version Compare
[unreleased]: 


cortex.py
#!/usr/bin/env python

import os
import pkg_resources
import click
from pycortexintelligence.core import funcs_startproject


VERSION = pkg_resources.require("pycortexintelligence")[0].version


@click.group()
@click.version_option(version=VERSION, prog_name='pycortex')
def pycortex():
    pass


@click.command()
@click.option('--name', help='Start a project with this name! Like a "Project Name"')
@click.option('--sname', help='Define a Safe Name to project like a "project_name"')
def startproject(name, sname):
    """
    This script create a Cortex Default Integration Template
    """
    # Validate args
    # Getting project Name
    if name is None:
        print('Example: Project Name')
        name = click.prompt('Give a Project Name.', type=str)
    if sname is None:
        print('Example: project_name')
        sname = click.prompt('Give a Project Safe Name.', type=str)
        if ' ' in sname:
            raise ValueError('Safe name, cannot have spaces!')

    # Call the functions
    print('Initializing tool...')
    path_real = os.path.dirname(os.path.realpath(__file__))
    path_fake = path_real.replace('/bin', "")
    funcs_startproject.check_create_dirs()
    funcs_startproject.creck_create_files(project_name=name, safe_name=sname)
    print('> Files are created on "{}"'.format(path_fake))
    print('Creating a Default Project Template for {}'.format(name))


pycortex.add_command(startproject)

if __name__ == '__main__':
    pycortex()

pycortexintelligence/core/funcs_startproject.py
import os
from pycortexintelligence.core.config import OUTPUT_FOLDER, DOWNLOADED_FOLDER

FOLDERS = [OUTPUT_FOLDER, DOWNLOADED_FOLDER]
FILES = [
    'README.md',
    '.gitignore',
    '.dockerignore',
    '.env',
    'requirements.txt',
    'Dockerfile',
    'main.py',
]


def write_main_py():
    return """from pycortexintelligence.core.funcs_env import load_env_to_dict, verify_env, delete_temp_files, check_dirs
from pycortexintelligence.core.config import OUTPUT_FOLDER, DOWNLOADED_FOLDER, LOG_FILE
from pycortexintelligence.core import messages
import logging

# Cria arquivo de LOG
logging.basicConfig(filename=LOG_FILE, encoding='utf-8', level=logging.DEBUG, format='%(asctime)s %(message)s')

# Set specific variables to project
custom_variables = ['variable_1', 'variable_2']

# Script para pegar automaticamente as variavéis de ambiente.
OS_PARAMS = load_env_to_dict(new_params_to_check=custom_variables)
logging.info(messages.VARIABLES_CREATED)

# Valida se o OS_PARAMS está correto para esse projeto.
verify_env(OS_PARAMS, new_params_to_check=custom_variables)
logging.info(messages.OS_PARAMS_CORRECT)

# Checa os diretórios temporários
check_dirs()
loggins.info(TEMP_DIRS)

# Deleta os arquivos temporários e de output.
delete_temp_files(os_params=OS_PARAMS)
logging.info(TEMP_DIRS_EMPTY)

# TODO Define a project!
"""


def write_dockerfile():
    return """FROM python:3.9-alpine

# Set arguments
# Arguments of E-mail
ENV email_password=""
ENV email_port=""
ENV email_smtp=""
ENV email_user=""
ENV email_to_error=""
ENV project_name=""

# Arguments of Cortex
ENV plataform_url=""
ENV plataform_username=""
ENV plataform_password=""
ENV plataform_cube_id=""

# Arguments of Project
ENV variable_1=""
ENV variable_2=""

# Set environment variables
ENV PYTHONUNBUFFERED 1
ENV PYTHONDONTWRITEBYTECODE 1

## install dependencies
RUN set -xe \\
    && apk add --no-cache curl \\
    && apk --no-cache add curl gcc g++ libressl-dev libffi-dev make \\
    && curl -sSL https://bootstrap.pypa.io/get-pip.py | python \\
    && pip install wheel

# Directory Structure
RUN mkdir -p /code

# Set work directory.
WORKDIR /code

# Copy project code.
COPY . /code/

# Install dependencies.
RUN pip install -r requirements.txt

CMD python main.py
"""


def write_env(project_name):
    return """email_password=""
email_port=""
email_smtp=""
email_user=""
email_to_error=""
project_name="{}"
plataform_url=""
plataform_username=""
plataform_password=""
plataform_cube_id=""
variable_1=""
variable_2=""
""".format(project_name)


def write_dockerignore():
    return """### Padrão Cortex ###
# Arquivos Temporárops e de Output
downloaded_files/*
output/*
historico.log
"""


def write_gitignore():
    return """# Created by https://www.toptal.com/developers/gitignore/api/python
# Edit at https://www.toptal.com/developers/gitignore?templates=python

### Padrão Cortex ###
# Arquivos Temporárops e de Output
downloaded_files/*
output/*
.env

### Python ###
# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class

# C extensions
*.so

# Distribution / packaging
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
pip-wheel-metadata/
share/python-wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST

# PyInstaller
#  Usually these files are written by a python script from a template
#  before PyInstaller builds the exe, so as to inject date/other infos into it.
*.manifest
*.spec

# Installer logs
pip-log.txt
pip-delete-this-directory.txt

# Unit test / coverage reports
htmlcov/
.tox/
.nox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
*.py,cover
.hypothesis/
.pytest_cache/
pytestdebug.log

# Translations
*.mo
*.pot

# Django stuff:
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal

# Flask stuff:
instance/
.webassets-cache

# Scrapy stuff:
.scrapy

# Sphinx documentation
docs/_build/
doc/_build/

# PyBuilder
target/

# Jupyter Notebook
.ipynb_checkpoints

# IPython
profile_default/
ipython_config.py

# pyenv
.python-version

# pipenv
#   According to pypa/pipenv#598, it is recommended to include Pipfile.lock in version control.
#   However, in case of collaboration, if having platform-specific dependencies or dependencies
#   having no cross-platform support, pipenv may install dependencies that don't work, or not
#   install all needed dependencies.
#Pipfile.lock

# PEP 582; used by e.g. github.com/David-OConnor/pyflow
__pypackages__/

# Celery stuff
celerybeat-schedule
celerybeat.pid

# SageMath parsed files
*.sage.py

# Environments
#.env
#.venv
#env/
#venv/
#ENV/
#env.bak/
#venv.bak/
#pythonenv*

# Spyder project settings
.spyderproject
.spyproject

# Rope project settings
.ropeproject

# mkdocs documentation
/site

# mypy
.mypy_cache/
.dmypy.json
dmypy.json

# Pyre type checker
.pyre/

# pytype static type analyzer
.pytype/

# profiling data
.prof

# End of https://www.toptal.com/developers/gitignore/api/python
"""


def write_readme_md(project_name, safe_name):
    return """# {}

**URL DO CONNECTOR:** `INSERIR URL DA TASK DENTRO DO ECS` 

**CUBO:** -

**FORM:** -

**PERIODICIDADE:** -

**CLIENTE:** -

**DATA CONNECTION:** `INSERIR URL DO CLUSTER DO ECS`

**DATA SOURCE:** `INSERIR URL DO ECR (Repositório GIT)`

**CARGA:** `CARGA REALIZADA VIA SID (PyCortexIntelligence)`

**CAMPOS ASSOCIADOS:**

`-` > `-`

**Descrição da Integração:**

- 

**Atenção**: 
 
- 

## How to test?

Create virtualenv and active it.

```shell
python -m venv {}
source {}/bin/activate
```

With the venv activated, you need install the requirements and run the project.
```
({}) pip install -r requirements.txt
({}) python main.py

```

## How to Build?

Build the container
```
docker build -t {} .
```

Run the container
```
docker run --rm {}
```
""".format(
        project_name,
        safe_name,
        safe_name,
        safe_name,
        safe_name,
        safe_name,
        safe_name,
    )


def check_create_dirs():
    for folder in FOLDERS:
        try:
            os.stat(folder)
        except FileNotFoundError:
            os.mkdir(folder)


def creck_create_files(project_name, safe_name):
    for file in FILES:
        with open(file, 'w', 'utf-8') as f:
            if file == 'README.md':
                file_to_write = write_readme_md(project_name, safe_name)
                print('Writing README.md file...')
            elif file == '.gitignore':
                file_to_write = write_gitignore()
                print('Writing .gitignore file...')
            elif file == '.dockerignore':
                file_to_write = write_dockerignore()
                print('Writing .dockerignore file...')
            elif file == '.env':
                file_to_write = write_env(project_name)
                print('Writing .env file...')
            elif file == 'requirements.txt':
                file_to_write = ""
                print('Writing requirements.txt file...')
            elif file == 'Dockerfile':
                file_to_write = write_dockerfile()
                print('Writing Dockerfile file...')
            elif file == 'main.py':
                file_to_write = write_main_py()
                print('Writing main.py file...')
            else:
                print('Unknown file.')
            f.writelines(file_to_write)


pycortexintelligence/functions.py
import datetime

import requests

from pycortexintelligence.core.messages import *

LOADMANAGER = "https://api.cortex-intelligence.com"


def _make_url_auth(plataform_url):
    return "https://{}/service/integration-authorization-service.login".format(plataform_url)


def _make_download_url(plataform_url):
    return 'https://{}/service/integration-cube-service.download?'.format(plataform_url)


def _get_sid_bearer_token(auth_endpoint, credentials):
    """
    :param auth_endpoint:
    :param credentials:
    :return:
    """
    response = requests.post(auth_endpoint, json=credentials)
    response_json = response.json()
    return {"Authorization": "Bearer " + response_json["key"]}


def _get_data_input(content, loadmanager, headers):
    """
    :param content:
    :param loadmanager:
    :param headers:
    :return:
    """
    endpoint = loadmanager + "/datainput"
    response = requests.post(endpoint, headers=headers, json=content)
    data_input_id = response.json()["id"]
    return data_input_id


def _get_execution_id(data_input_id, content, loadmanager, headers):
    """
    :param data_input_id:
    :param content:
    :param loadmanager:
    :param headers:
    :return:
    """
    endpoint = "{}/datainput/{}/execution".format(loadmanager, data_input_id)
    response = requests.post(endpoint, headers=headers, json=content)
    execution_id = response.json()["executionId"]
    return execution_id


def _start_process(execution_id, loadmanager, headers):
    """
    :param execution_id:
    :param loadmanager:
    :param headers:
    :return:
    """
    endpoint = loadmanager + "/execution/" + execution_id + "/start"
    response = requests.put(endpoint, headers=headers)
    return response


def _execution_history(execution_id, loadmanager, headers):
    """
    :param execution_id:
    :param loadmanager:
    :param headers:
    :return:
    """
    endpoint = loadmanager + "/execution/" + execution_id  # + '/history'
    response = requests.get(endpoint, headers=headers)
    return response


def upload_local_2_cube(cubo_id,
                        file_path,
                        auth_endpoint,
                        credentials,
                        loadmanager=LOADMANAGER,
                        data_format={
                            "charset": "UTF-8",
                            "quote": "\"",
                            "escape": "\\",
                            "delimiter": ",",
                            "fileType": "CSV"
                        },
                        timeout={
                            'file': 300,
                            'execution': 600,
                        },
                        execution_parameters={
                            'name': 'LoadManager PyCortex',
                        },
                        datainput_parameters={
                            'ignoreValidationErrors': False
                        }
                        ):
    """
    :param timeout:
    :param cubo_id:
    :param file_path:
    :param auth_endpoint:
    :param credentials:
    :param loadmanager:
    :param data_format:
    :return:
    """

    # ================ Get Bearer Token ===================
    headers = _get_sid_bearer_token(auth_endpoint, credentials)

    # ================ Content ============================
    content = {
        "destinationId": cubo_id,
        'ignoreValidationErrors': datainput_parameters['ignoreValidationErrors'],
        "fileProcessingTimeout": int(timeout['file']),
        "executionTimeout": int(timeout['execution']),
    }

    # ================ Get Data Input Id ======================
    data_input_id = _get_data_input(content, loadmanager, headers)

    # ================ Get Execution Id =======================
    execution_id = _get_execution_id(data_input_id, execution_parameters, loadmanager, headers)

    # ================ Send files =============================
    endpoint = loadmanager + "/execution/" + execution_id + "/file"
    response = requests.post(
        endpoint,
        headers=headers,
        data=data_format,
        files={"file": open(file_path, "rb")},
    )

    # ================ Start Data Input Process ===========================
    _start_process(execution_id, loadmanager, headers)

    return execution_id, headers


def upload_to_cortex(**kwargs):
    """
    :param cubo_id:
    :param file_path:
    :param plataform_url:
    :param username:
    :param password:
    :param data_format: data_format={
                            "charset": "UTF-8",
                            "quote": "\"",
                            "escape": "\\",
                            "delimiter": ",",
                            "fileType": "CSV"
                        }
    :param timeout: {
        'file': 300,
        'execution': 600,
    }
    :param loadmanager: LOADMANAGER
    :return:
    """
    # Read Kwargs
    cubo_id = kwargs.get('cubo_id')
    file_path = kwargs.get('file_path')
    plataform_url = kwargs.get('plataform_url')
    username = kwargs.get('username')
    password = kwargs.get('password')
    data_format = kwargs.get('data_format', {
        "charset": "UTF-8",
        "quote": "\"",
        "escape": "\\",
        "delimiter": ",",
        "fileType": "CSV"
    })
    timeout = kwargs.get('timeout', {
        'file': 300,
        'execution': 600,
    })
    loadmanager = kwargs.get('loadmanager', LOADMANAGER)
    execution_parameters = kwargs.get('execution_parameters', {
        'name': 'LoadManager PyCortex',
    })
    datainput_parameters = kwargs.get('datainput_parameters', {
        'ignoreValidationErrors': False
    })

    if 'file' not in timeout.keys() and 'execution' not in timeout.keys():
        raise ValueError(FORMAT_TIMEOUT)

    # Verify Kwargs
    if cubo_id and file_path and plataform_url and username and password:
        auth_endpoint = _make_url_auth(plataform_url)
        credentials = {"login": str(username), "password": str(password)}
        execution_id, headers = upload_local_2_cube(
            cubo_id=cubo_id,
            file_path=file_path,
            auth_endpoint=auth_endpoint,
            credentials=credentials,
            data_format=data_format,
            timeout=timeout,
            loadmanager=loadmanager,
            execution_parameters=execution_parameters,
            datainput_parameters=datainput_parameters,
        )
        response = _execution_history(execution_id, LOADMANAGER, headers)
        return response
    else:
        raise ValueError(ERROR_ARGUMENTS_VALIDATION)


def download_from_cortex(**kwargs):
    """
    :param cubo_id:
    :param cubo_name:
    :param plataform_url:
    :param username:
    :param password:
    :param columns:
    :param file_path:
    :param data_format:
    :param filters:
    :return:
    """
    cubo_id = kwargs.get('cubo_id')
    cubo_name = kwargs.get('cubo_name')
    plataform_url = kwargs.get('plataform_url')
    username = kwargs.get('username')
    password = kwargs.get('password')
    columns = kwargs.get('columns')
    file_path = kwargs.get('file_path')
    data_format = kwargs.get('data_format', {
        "charset": "UTF-8",
        "quote": "\"",
        "escape": "\\",
        "delimiter": ",",
    })
    filters = kwargs.get('filters', None)
    if cubo_id and cubo_name:
        raise ValueError(DOWNLOAD_ERROR_JUST_ID_OR_NAME)
    if (cubo_id or cubo_name) and plataform_url and username and password and columns and file_path:
        # Verify is a ID or Name
        if cubo_id:
            cube = '{"id":"' + cubo_id + '"}'
        else:
            cube = '{"name":"' + cubo_name + '"}'

        # Columns to Download
        columns_download = []
        for column in columns:
            columns_download.append({
                "name": column,
            })
        columns_download = str(columns_download).replace("'", '"')

        # Need to Apply Filters
        if filters:
            filters_download = []
            for filter in filters:
                column_name = filter[0]
                value = filter[1]
                element = {
                    "name": column_name,
                    "type": "SIMPLE",
                }
                try:
                    value = datetime.datetime.strptime(value, "%d/%m/%Y")
                    element["type"] = "DATE"
                    element["rangeStart"] = value.strftime("%Y%m%d")
                    element["rangeEnd"] = value.strftime("%Y%m%d")
                except ValueError:
                    value_temp = value
                    try:
                        value = value.split('-')
                        date_start = datetime.datetime.strptime(value[0], "%d/%m/%Y")
                        date_end = datetime.datetime.strptime(value[1], "%d/%m/%Y")
                        element["type"] = "DATE"
                        element["rangeStart"] = date_start.strftime("%Y%m%d")
                        element["rangeEnd"] = date_end.strftime("%Y%m%d")
                    except ValueError:
                        value = value_temp.split('|')
                        element["value"] = value
                filters_download.append(element)
            filters_download = str(filters_download).replace("'", '"')

        auth_endpoint = _make_url_auth(plataform_url)
        credentials = {"login": str(username), "password": str(password)}
        auth_post = requests.post(auth_endpoint, json=credentials)
        headers = {
            'x-authorization-user-id': auth_post.json()['userId'],
            'x-authorization-token': auth_post.json()['key']
        }
        download_endpoint = _make_download_url(plataform_url)
        payload = {
            'cube': cube,
            'charset': data_format['charset'],
            'delimiter': data_format['delimiter'],
            'quote': data_format['quote'],
            'escape': data_format['escape'],
        }
        if filters:
            payload['filters'] = filters_download
        if columns_download:
            payload['headers'] = columns_download

        with requests.get(download_endpoint, stream=True, headers=headers, params=payload) as r:
            r.raise_for_status()
            with open(file_path, 'wb') as f:
                for chunk in r.iter_content(chunk_size=8192):
                    f.write(chunk)
    else:
        raise ValueError(ERROR_ARGUMENTS_VALIDATION)

